lib/puppetserver/ca/revoke_action.rb

```diff
@@ -0,0 +1,149 @@
+HTTPS using the local agent's PKI
+
+Options:
+BANNER
+
+      def self.parser(parsed = {})
+        OptionParser.new do |o|
+          o.banner = BANNER
+          o.on('--config', 'Path to puppet.conf') do |conf|
+            parsed['config'] = conf
+          end
+        end
+      end
+
+      def initialize(logger)
+        @logger = logger
+      end
+
+      def parse(args)
+        results = {}
+        parser = self.class.parser(results)
+
+        _, certnames, invalid = Utils.parse_without_raising(parser, args)
+
+        exit_code = nil
+        unless invalid.empty?
+          Utils.print_invalid_args(@logger, invalid, parser.help)
+          exit_code = 1
+        end
+
+        case certnames.length
+        when 0
+          @logger.err 'A certname to revoke is required'
+          @logger.err parser.help
+          exit_code = 1
+        when 1
+          results[:certname] = certnames.first
+        else
+          @logger.err 'Only one certname is allowed at a time'
+          @logger.err 'Received certnames:'
+          certnames.each {|name| @logger.err "    #{name}" }
+          exit_code = 1
+        end
+
+        return results, exit_code
+      end
+
+      # take the parsed arg hash from #parse and perform this action
+      # print information to @logger and return a valid exit code
+      def run(args)
+        certname = args[:certname]
+        config = args[:config]
+
+        errors = validate_file_paths(config)
+        unless errors.empty?
+          @logger.err 'Error:'
+          errors.each {|error| @logger.err "    #{error}" }
+          exit 1
+        end
+
+        puppet = PuppetConfig.parse(config)
+
+        revoke(certname, puppet.settings)
+      end
+
+      def make_store(cacert, cert, crl)
+        store = OpenSSL::X509::Store.new
+        store.add_file(cacert)
+        store.add_file(cert)
+        store.add_file(crl)
+
+        store
+      end
+
+      def make_uri(host, port, certname)
+        URI('https://' + host + ':' + port +
+            '/puppet-ca/v1/certificate_status/' +
+            certname +
+            '?environment=production')
+      end
+
+      def revoke(certname, settings)
+        body = JSON.dump({ desired_state: 'revoked' })
+
+        headers = {
+          'User-Agent' => 'PuppetserverCaCli',
+          'Content-Type' => 'application/json',
+          'Accept' => 'application/json'
+        }
+
+        store = make_store(settings[:localcacert],
+                           settings[:hostcert],
+                           settings[:hostcrl])
+
+        uri = make_uri(settings[:ca_server],
+                       settings[:ca_port],
+                       certname)
+
+        result = Net::HTTP.start(uri.host, uri.port, use_ssl: true, cert_store: store) do |connection|
+          request = Net::HTTP::Put.new(uri, headers)
+          request.body = body
+          connection.request(request)
+        end
+
+        if result.is_a? Net::HTTPSuccess
```

Comment:
both cert and the api will gladly return success if the cert has already been revoked, can we fix that in the api as part of this?

Reply:
Probably. What response would you expect?

Reply:
Hrm... I hadn't thought about the specifics. We're doing a PUT which is supposed to be idempotent (and we retain the list of revoked certs in the inventory), so theoretically I suppose this should be a success as long as the end state of the http call is that the cert is revoked regardless of its state beforehand.

Maybe instead I should make two calls? One to query the state of the cert being revoked and then one to update it?

Reply:
This is starting to sound even more like @lazartea 's PR. An `--all` flag doesn't make sense for revoke (it does/did exist in `puppet cert` and it been a feature request to _remove_ that ability for years), but should we accept a `--certname` flag with that potentially contains a list of comma separated certnames?

Reply:
I think it doesn't work now. The docs say you can use it, but if you use it, it errors. At least that's the way it works for `clean`.

I like the idea of accepting a `--certname` flag here. Consistency seems good.

Reply:
I'm also fine returning success so long as the cert ends up revoked, whether or not it started out that way.